rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+
+
+##### `dataset_fn` path
+
+In TF2 parameter server training, `ClusterCoordinator` naturally supports a dataset function to be passed in to `create_per_worker_dataset` API, which creates datasets on remote workers. By leveraging such data factory support, `model.fit` with `dataset_fn` can be implemented by subclassing the existing Keras `DataHandler` to provide a worker-distributed dataset for Keras to use (i.e. call `iter` on). Please see `DataHandler` section below for proposed changes.
+
+In terms of how users pass a dataset factory into `model.fit`, there are a couple of options:
+
+
+###### Option 1: any `callable`
+
+In the simplest case, we can allow any kind of `callable` to be passed in:
+
+
+```
+def dataset_fn(): 
+  return tf.data.Dataset.from_tensor_slices(...)
+history = model.fit(dataset_fn, epochs=..., steps_per_epoch=...,  callbacks=[...])
+```
+
+
+
+###### Option 2: dataset factory
+
+For future-compatibility of `model.fit` API where a `dataset_fn` may have a signature change, a `DatasetFactory` can come handy which determines how the function is supposed to be used. In the case where a `ClusterCoordinator` is used, it is supposed to be invoked with no arguments. With this, user flow will become:
+
+
+```
+def dataset_fn(): return tf.data.Dataset.from_tensor_slices(...)
+history = model.fit(DatasetFactory(dataset_fn), epochs=..., steps_per_epoch=...,  callbacks=[...])
+```
+
+
+With an additionally defined class `DatasetFactory`:
+
+
+```
+class DataFactory(object):
+
+  def __init__(self, x):
+    if not callable(x):
+      raise TypeError('Input for `DataFactory` must be a `callable`.')
+    self.x = x
+
+  def __call__(self, *args, **kwargs):
+    # We gain the flexibility of modifying args/kwargs for future-compatibility.
+    # For example, if we allow different argument signature of user-provided
+    # `dataset_fn`, this works as an abstraction layer. 
+    # If we now only allow zero-arg, but later extend it to one-arg (e.g. input
+    # context), we omit the the arg when we observe that the function doesn't
+    # take any arg.
+    return self.x(*args, **kwargs)
+```
+
+
+The following discussion is tentatively based on option 1, where a simple callable is taken.
+
+
+###### Implication on no strategy/other strategies
+
+If `model.fit` is allowed to take a `dataset_fn`, use cases for synchronous strategies, and no strategy, can be readily applied. That is, when a dataset is needed, the `callable` is invoked eagerly to obtain a `dataset`, and the remaining workflow will apply.
+
+
+###### Signature of `dataset_fn`
+
+The signature (input argument and return value) of `dataset_fn` taken by `model.fit` should basically follow the signature `ClusterCoordinator.create_per_worker_dataset` takes. There has been discussion around whether that should take an `InputContext` for effective sharding. Current decision is that it does not, since we do not expect users to shard the dataset considering workers can get preempted. 
```

Comment:
Good call. This should not be user-facing

Reply:
Correct